### PR TITLE
[WIP] Fixed #27704 -- Used TypedMultipleChoiceField for array fields with choices

### PR DESCRIPTION
**Pull request by claudep**
https://code.djangoproject.com/ticket/27704

### CODE REVIEW

File: tests/postgres_tests/test_array.py [hunk 1]
```diff
@@ -786,3 +786,22 @@ def test_value_omitted_from_data(self):
         self.assertIs(widget.value_omitted_from_data({'field_0': 'value'}, {}, 'field'), False)
         self.assertIs(widget.value_omitted_from_data({'field_1': 'value'}, {}, 'field'), False)
         self.assertIs(widget.value_omitted_from_data({'field_0': 'value', 'field_1': 'value'}, {}, 'field'), False)
+
+
+class TestChoiceFormField(PostgreSQLTestCase):
+    def test_model_field_formfield(self):
+        model_field = ArrayField(models.CharField(max_length=27), choices=(('a1', 'A1'), ('b1', 'B1')))
```
Comment (rtnpro) [inline tests/postgres_tests/test_array.py]: I tried saving this model in from admin page, I get the following validation error:

``Select a valid choice. ['a1', 'b1'] is not one of the available choices.``

**Comment by timgraham** [comment]
https://github.com/django/django/pull/8358 passes the tests in this PR (except for the one that's checking for TypedMultipleChoiceField). It looks like a simpler approach -- any concerns with it?

**Comment by claudep** [comment]
I don't mind committing #8538, but  that PR doesn't address the issue of the `TypedMultipleChoiceField` usage, of course.

**Comment by timgraham** [comment]
Hopefully this patch can now be simplified a bit after 9dd244394236388c3479ab202a0ec31055f7ec09.

**Comment by LuRsT** [comment]
@claudep are you still working on this one? Should it stay open?

**Comment by felixxm** [comment]
@LuRsT It is WIP so it can stay open, that doesn't hurt. It is not really helpful to leave comments/pings in dozens on PRs. Would you like to help with all of them? I would suggest to focus on few PRs and really help to move them forward.

**Comment by LuRsT** [comment]
Fair enough @felixxm, I was just trying to keep the number of open PRs down if they are not being currently open, I'm happy not to go through them to nudge people if that's not useful. Is it OK to pick someone's PR and change it? Or even create a new one based off of someones PR?

**Comment by felixxm** [comment]
@LuRsT Yes that's OK. You can supersed someone PR if it is inactive for a while.

**Comment by felixxm** [comment]
Closing due to inactivity.

**Comment by pope1ni** [comment]
@claudep I was thinking about this a little bit more to work out how we can get this resolved and I had some questions.

Can we think of a good reason why we should avoid using `TypedMultipleChoiceField` with `choices` in an `ArrayField`? The approach that was started here seems to imply that `SimpleArrayField` would be used if `choices` were defined on the `ArrayField` directly, and `TypedMultipleChoiceField` would be used if `choices` were defined on the `base_field`. The first commit in this message added some documentation to that effect, but it feels to me that this could get confusing.

In fact it turns out that `SimpleArrayField` is never used when `ArrayField.choices` is defined. In that case the default value of `choices_form_class` is used in `Field.formfield()` which happens to be `TypeChoiceField` -- a poor fit for an array field. So firstly I think that we should _always_ set `choices_form_class=TypedMultipleChoiceField` so that this works wherever we put `choices`.

ISTM, however, that we should prevent `choices` being defined on the `ArrayField` and force it to be used on the `base_field` instead to avoid confusion. We could make `ArrayField.choices` push down to `base_field.choices` and have a system check warning too. We could also deprecate use of `ArrayField.choices` and remove support for it entirely eventually?

**Comment by claudep** [comment]
Sorry Nick, I'm a bit overloaded those days. I'll not be able to dig again into it in the short term. Your plan sounds good at first sight, please go ahead!

**Pull request closed without merging**